File: C/416.partition-equal-subset-sum.c
/*
 * @lc app=leetcode id=416 lang=c
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
int cmp(const void *a, const void *b){
    return *(int *)a < *(int *)b;
}

int sum(int* nums, int numsSize){
    int s = 0;
    for(int i = 0; i < numsSize; i ++){
        s += nums[i];
    }
    return s;
}
int findmax(int* nums, int numsSize){
    int max = 0;
    for(int i = 0; i < numsSize; i ++){
        if(max < nums[i]) max = nums[i];
    }
    return max;
}

int dfs(int target, int curpos, int* nums, int numsSize){
    if(target < 0) return 0;
    if(target == 0) return 1;
    for(int i = curpos; i < numsSize; i ++){
        if(dfs(target - nums[i], i + 1, nums, numsSize) == 1) return 1;
    }
    return 0;
}
bool canPartition(int* nums, int numsSize){
    int s = sum(nums, numsSize);
    if(s % 2 != 0) return 0;
    int max_num = findmax(nums, numsSize);
    int target = s >> 1;
    if(max_num > target) return 0;
    qsort(nums, numsSize, sizeof(int), cmp);
    return dfs(target, 0, nums, numsSize);

}


// @lc code=end

File: Python3/416.partition-equal-subset-sum.py
#
# @lc app=leetcode id=416 lang=python3
#
# [416] Partition Equal Subset Sum
#

# @lc code=start

# non-empty array positive integers only - we don't need to deal with [] input
# can it partitioned into two subset with equal sum? return True if can, False if can't

class Solution:
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :return: bool
        """
        s = sum(nums)
        # if sum is an odd number
        if s & 1:
            return False
        target = s//2
        if max(nums) > target:
            return False
        nums.sort(reverse = True)
        return self.dfs(nums, 0, target)
        
    def dfs(self, nums, idx, target):
        if target == 0:
            return True
        if target < 0:
            return
        for i in range(idx, len(nums)):
            if self.dfs(nums, i + 1, target - nums[i]):
                return True
        return False

# class Solution:
#     def canPartition(self, nums):
#         """
#         :type nums: List[int]
#         :rtype: bool
#         """
#         def dfs(start, target):
#             if target < 0:
#                 return
#             if target == 0:
#                 return True
#             for i in range(start, len(nums)):
#                 if dfs(i+1, target - nums[i]):
#                     return True
#             return False
            
#         s = sum(nums)
#         if s % 2 != 0:
#             return False
        
#         nums.sort(reverse = True)
#         return dfs(0, s//2)

a = Solution()
b = a.canPartition([2, 2, 1, 1])
print(b)

# class Solution:
#     def canPartition(self, nums: List[int]) -> bool:

#         #Time Complexity: O(n)
#         #Space Complexity: O(n)

#         Sum = sum(nums)

#         if Sum % 2 == 1:
#             return False

#         Sum = Sum//2

#         n = len(nums)

#         dp = [0]*(Sum+1)
#         dp[0] = 1

#         for i in range(1, n+1):
#             dp_new = dp.copy()

#             for j in range(1, Sum+1):
#                 if j - nums[i-1] >= 0 and dp[j-nums[i-1]]:
#                     dp_new[j] = 1
#             dp = dp_new


#         return dp[-1]

# @lc code=end

